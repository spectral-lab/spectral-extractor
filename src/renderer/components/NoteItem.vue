<template>
  <g>
    <g v-if="isEdited">
      <g
        v-for="(point, idx) in positions"
        :key="'a' + point.id"
        class="note-action-group"
      >
        <line
          v-if="nextPoint(idx)"
          :key="'c' + point.id"
          :x1="`${point.x * 100}%`"
          :y1="`${point.y * 100}%`"
          :x2="`${nextPoint(idx).x * 100}%`"
          :y2="`${nextPoint(idx).y * 100}%`"
          :stroke="lineColor"
          :stroke-width="2"
          :opacity="1"
        />
        <circle
          :cx="`${point.x * 100}%`"
          :cy="`${point.y * 100}%`"
          :fill="circleColor"
          :opacity="0.8"
          r="5"
        />
      </g>
    </g>
    <g v-else>
      <g
        v-for="(point, idx) in positions"
        :key="'d' + point.id"
        class="note-line"
      >
        <line
          v-if="nextPoint(idx)"
          :key="'f' + point.id"
          :x1="`${point.x * 100}%`"
          :y1="`${point.y * 100}%`"
          :x2="`${nextPoint(idx).x * 100}%`"
          :y2="`${nextPoint(idx).y * 100}%`"
          :stroke="lineColor"
          :stroke-width="10"
          :opacity="1"
          @click="handleClick"
          @dblclick="handleDblClick"
        />
        <circle
          :cx="`${point.x * 100}%`"
          :cy="`${point.y * 100}%`"
          :fill="circleColor"
          :opacity="1"
          @click="handleClick"
          @dblclick="handleDblClick"
          r="5"
        />
      </g>
    </g>
  </g>
</template>

<script>
import { tickToPosX, pitchToPosY } from '../modules/pianoRoll/utils';

export default {
  props: {
    noteId: String,
    pitchTransition: Array,
    color: {
      type: String,
      default: 'grey'
    },
    totalTicks: Number,
    isSelected: {
      type: Boolean,
      default: false
    },
    isEdited: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    positions () {
      return this.pitchTransition.map(point => ({
        x: tickToPosX(point.offsetTime, this.totalTicks),
        y: pitchToPosY(point.pitch),
        id: point.id,
        type: point.type
      }));
    },
    lineColor () {
      if (this.isEdited) return 'Honeydew';
      if (this.isSelected) return 'Honeydew';
      return this.color;
    },
    circleColor () {
      if (this.isEdited) return this.color;
      if (this.isSelected) return 'Honeydew';
      return this.color;
    }
  },
  methods: {
    nextPoint (idx) {
      return this.positions[idx + 1];
    },
    handleClick (ev) {
      this.$emit('click', ev, this.noteId);
    },
    handleDblClick (ev) {
      this.$emit('dblclick', ev, this.noteId);
    }
  }
};
</script>

<style scoped>
</style>
