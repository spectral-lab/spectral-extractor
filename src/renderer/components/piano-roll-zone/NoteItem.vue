<template>
  <g>
    <defs>
      <filter id="shadow">
        <feDropShadow
          dx="0.2"
          dy="0.4"
          std-deviation="2"
        />
      </filter>
    </defs>
    <g
      v-if="isEdited"
    >
      <g
        v-for="(point, idx) in positions"
        :key="'a' + point.id"
        class="note-action-group"
      >
        <line
          v-if="nextPoint(idx)"
          :key="'c' + point.id"
          :x1="`${point.x * 100}%`"
          :y1="`${point.y * 100}%`"
          :x2="`${nextPoint(idx).x * 100}%`"
          :y2="`${nextPoint(idx).y * 100}%`"
          :stroke="lineColor"
          :stroke-width="2"
        />
        <circle
          :cx="`${point.x * 100}%`"
          :cy="`${point.y * 100}%`"
          :fill="circleColor"
          r="5"
          style="filter:url(#shadow);"
        />
      </g>
    </g>
    <g
      v-else
      style="filter:url(#shadow);"
    >
      <g
        v-for="(point, idx) in positions"
        :key="'d' + point.id"
        class="note-line"
      >
        <line
          v-if="nextPoint(idx)"
          :key="'f' + point.id"
          :x1="`${point.x * 100}%`"
          :y1="`${point.y * 100}%`"
          :x2="`${nextPoint(idx).x * 100}%`"
          :y2="`${nextPoint(idx).y * 100}%`"
          :stroke="lineColor"
          :stroke-width="10"
          @click="handleClick"
          @dblclick="handleDblClick"
        />
        <circle
          :cx="`${point.x * 100}%`"
          :cy="`${point.y * 100}%`"
          :fill="circleColor"
          @click="handleClick"
          @dblclick="handleDblClick"
          r="5"
        />
      </g>
    </g>
  </g>
</template>

<script>
import { tickToPosX, pitchToPosY } from '../../utils/view/pianoRoll/utils';

export default {
  props: {
    noteId: {
      type: String,
      default: 'no_id'
    },
    pitchTransition: {
      type: Array,
      default: () => []
    },
    color: {
      type: String,
      default: 'grey'
    },
    totalTicks: {
      type: Number,
      default: 4800
    },
    isSelected: {
      type: Boolean,
      default: false
    },
    isEdited: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    positions () {
      return this.pitchTransition.map(point => ({
        x: tickToPosX(point.offsetTime, this.totalTicks),
        y: pitchToPosY(point.pitch),
        id: point.id,
        type: point.type
      }));
    },
    lineColor () {
      if (this.isEdited) return 'Honeydew';
      if (this.isSelected) return 'Honeydew';
      return this.color;
    },
    circleColor () {
      if (this.isEdited) return this.color;
      if (this.isSelected) return 'Honeydew';
      return this.color;
    }
  },
  methods: {
    nextPoint (idx) {
      return this.positions[idx + 1];
    },
    handleClick (ev) {
      this.$emit('click', ev, this.noteId);
    },
    handleDblClick (ev) {
      this.$emit('dblclick', ev, this.noteId);
    }
  }
};
</script>

<style scoped>
</style>
