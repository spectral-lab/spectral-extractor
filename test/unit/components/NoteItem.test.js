import { shallowMount } from '@vue/test-utils';
import NoteItem from '../../../src/renderer/components/NoteItem';
import mockPitchTransition from '../../data/mockPitchTransition';

describe('Component', () => {
  test('is a Vue instance', () => {
    const wrapper = shallowMount(NoteItem, {
      propsData: {
        pitchTransition: mockPitchTransition,
        totalTicks: 10000
      }
    });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });
  test('renders correctly', () => {
    const wrapper = shallowMount(NoteItem, {
      propsData: {
        pitchTransition: mockPitchTransition,
        totalTicks: 10000
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
