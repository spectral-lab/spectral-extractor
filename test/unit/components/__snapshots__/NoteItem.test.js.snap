// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders correctly 1`] = `
<g>
  <defs>
    <filter
      id="shadow"
    >
      <fedropshadow
        dx="0.2"
        dy="0.4"
        std-deviation="2"
      />
    </filter>
  </defs>
   
  <g>
    <g
      class="note-line"
    >
      <line
        stroke="grey"
        stroke-width="10"
        x1="3.048680272108843%"
        x2="5.260136054421769%"
        y1="45.643767533234914%"
        y2="69.08126753323438%"
      />
       
      <circle
        cx="3.048680272108843%"
        cy="45.643767533234914%"
        fill="grey"
        r="5"
      />
    </g>
    <g
      class="note-line"
    >
      <line
        stroke="grey"
        stroke-width="10"
        x1="5.260136054421769%"
        x2="7.371591836734695%"
        y1="69.08126753323438%"
        y2="22.206267533234115%"
      />
       
      <circle
        cx="5.260136054421769%"
        cy="69.08126753323438%"
        fill="grey"
        r="5"
      />
    </g>
    <g
      class="note-line"
    >
      <line
        stroke="grey"
        stroke-width="10"
        x1="7.371591836734695%"
        x2="12.48304761904762%"
        y1="22.206267533234115%"
        y2="14.39376753323267%"
      />
       
      <circle
        cx="7.371591836734695%"
        cy="22.206267533234115%"
        fill="grey"
        r="5"
      />
    </g>
    <g
      class="note-line"
    >
      <!---->
       
      <circle
        cx="12.48304761904762%"
        cy="14.39376753323267%"
        fill="grey"
        r="5"
      />
    </g>
  </g>
</g>
`;
